From af8e9cc26ba7aa56b4440edf3aa9792db4daa1d0 Mon Sep 17 00:00:00 2001
From: LuK1337 <priv.luk@gmail.com>
Date: Thu, 25 Jan 2018 13:33:58 +0100
Subject: [PATCH] Fix division by zero exception when changing volume

* For some reason HeadsetClientStateMachine isn't
  always initialized and thus calling amToHfVol()
  throws an exception due to uninitialized voice
  call stream volumes.

Change-Id: Iffb9fc7c03a2f8211164135a746ef45f0e88dd7a
---
 .../bluetooth/hfpclient/HeadsetClientService.java     | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)

diff --git a/src/com/android/bluetooth/hfpclient/HeadsetClientService.java b/src/com/android/bluetooth/hfpclient/HeadsetClientService.java
index 089f4a6..9530106 100644
--- a/src/com/android/bluetooth/hfpclient/HeadsetClientService.java
+++ b/src/com/android/bluetooth/hfpclient/HeadsetClientService.java
@@ -171,20 +171,25 @@ public class HeadsetClientService extends ProfileService {
                 if (streamType == AudioManager.STREAM_VOICE_CALL) {
                     int streamValue = intent
                             .getIntExtra(AudioManager.EXTRA_VOLUME_STREAM_VALUE, -1);
-                    int hfVol = HeadsetClientStateMachine.amToHfVol(streamValue);
-                    if (DBG) {
-                        Log.d(TAG,
-                                "Setting volume to audio manager: " + streamValue
-                                        + " hands free: " + hfVol);
-                    }
-                    mAudioManager.setParameters("hfp_volume=" + hfVol);
                     synchronized (this) {
+                        int hfVol = -1;
                         for (HeadsetClientStateMachine sm : mStateMachineMap.values()) {
                             if (sm != null) {
+                                if (hfVol == -1) {
+                                    hfVol = sm.amToHfVol(streamValue);
+                                }
                                 sm.sendMessage(
                                         HeadsetClientStateMachine.SET_SPEAKER_VOLUME, streamValue);
                             }
                         }
+                        if (hfVol != -1) {
+                            if (DBG) {
+                                Log.d(TAG,
+                                        "Setting volume to audio manager: " + streamValue
+                                                + " hands free: " + hfVol);
+                            }
+                            mAudioManager.setParameters("hfp_volume=" + hfVol);
+                        }
                     }
                 }
             }
-- 
2.7.4

